### ISSUE TITLE
Rich feature hierarchies for accurate object detection and semantic segmentation

### ISSUE THREAD

**Issue by gimmeyoon** [gimmeyoon/all-the-deep-learning#1]
First paper reading for my new project, "Instance Segmentation"

ref. [link](https://www.cv-foundation.org/openaccess/content_cvpr_2014/papers/Girshick_Rich_Feature_Hierarchies_2014_CVPR_paper.pdf)

**Comment by gimmeyoon** [comment]
## Abstract

* Object detection on PASCAL VOC dataset: complex ensemble systems that typically combine multiple low-level image features with high-level context.
* In this paper
  * **mAP of 53.3%**
  * R-CNN == **Regions with CNN features**
  * Apply high-capacity CNNs to bottom-up region proposals in order to localize and segment objects
  * (if labeled training data is scarce) pre-training + domain-specific fine-tuning

## Introduction

* Visual recognition tasks has been based considerably on: SIFT and HOG
  * (PASCAL VOC object detection) It has been slow during 2010-2012
* After image classification with CNNs
  * **To what extent do the CNN classification results on ImageNet generalize to object detection results on the PASCAL VOC Challenge?**
    * Bridging the chasm between image classification and object detection.
* Show that a CNN can lead to dramatically higher object detection performance on PASCAL
VOC as compared to systems based on simpler HOG-like features
  * **Localizing objects with a deep network**
  * **Training a high-capacity model with only a small quantity of annotated detection data**

![image](https://user-images.githubusercontent.com/46519736/71402736-fd889200-2670-11ea-9da0-4693adfdbd59.png)

### Localization

* Unlike image classification, detection requires localizing objects within an image
  * :x: Regression problem?
  * :x: Sliding-Window?
  * :o: **Recognition using regions**

### Unsupervised manner

* Labeled data is scarce and the amount currently available is insufficient for training a large CNN
  * Solution: **Use unsupervised pre-training, followed by supervised fine-tuning**
* Supervised pretraining on a large auxiliary dataset (ILSVRC), followed by domain-specific fine-tuning on a small dataset (PASCAL), is an effective paradigm for learning high-capacity CNNs when data is scarce
  *  In this experiments, fine-tuning for detection improves mAP performance by 8 percentage points

## Object detection with R-CNN

* (1) Category-independent region proposals.
* (2) Large convolutional neural network that extracts a fixed-length feature
* (3) A set of classspecific linear SVMs.

### Category-independent region proposals

* A variety of recent papers offer methods for generating category-independent region proposals.
  * [1], [32], [11], [5], [3], [6]
* While R-CNN is agnostic to the particular region proposal method, we use **selective search** to enable a controlled comparison with prior

### CNNs

* Extract a 4096-dimensional feature vector from each region proposal using the Caffe [21]
implementation of the CNN described by Krizhevsky et al. [22]
  * Features are computed by forward propagating a mean-subtracted 227 × 227 RGB image through five convolutional layers and two fully connected layers
* In order to compute features for a region proposal,
  * Regardless of the size or aspect ratio of the candidate region, we warp all pixels in a tight bounding box around it to the required size.
  * Prior to warping, we dilate the tight bounding box so that at the warped size there are exactly p pixels of warped image context around the original box (we use p = 16).

**Comment by gimmeyoon** [comment]
## Visualization, ablation, and modes of error

### Visualizing learned features

* Non-parametric method that shows what the network learned
  * Single out a particular unit in the network/use it as if it were an object detector in its own right
* Visualize units from layer pool_5
  * max-pooled output of the network's 5th and final convolutional layer
  * 6 x 6 x 256 = 9,216-dim
* **Can't understand**

>Ignoring boundary effects, each pool_5 unit has a receptive field of 195 x 195 pixels in the original 227 x 227 pixel input. A central pool_5 unit has a nearly global view, while one near the edge has a smaller, clipped support.

![image](https://user-images.githubusercontent.com/46519736/71569011-37451700-2b0f-11ea-9a83-08a3e7d875d6.png)

* The network appears to learn a representation that combines a small number of class-tuned features together with a distributed representation of shape, texture, color, and material properties
  * The subsequent fully connected layer fc_6 has the ability to model a large set of compositions of these rich features

### Ablation studeis

![image](https://user-images.githubusercontent.com/46519736/71569508-f3541100-2b12-11ea-9a36-e3520b0758f3.png)

#### Performance layer-by-layer without fine-tuning

* +fc_7 < -fc_7: 29%, or about 16.8m of the CNN's parameters can be removed without degrading mAP
* -fc_7 - fc_6 quite good: features are computed using only 6% of the CNN's parameters
* Much of the CNN's representational power comes from its convolutional layers

#### Performance layer-by-layer with fine-tuning

* The boos from fine-tuning is much larger for fc_6 and fc_7 than for pool_5
  * Features learned from ImageNet are general and that most of the improvement is gained from learning domain-specific non-linear classifiers on top of them

## Semantic Segmentation

* **full**: Ignores the region's shape and computes CNN features directly on the warped window
  * These features ignore the non-rectangular shape of the region
  * Two regions might have very similar bounding boxes while having very little overlap
* **fg**: Computes CNN features only on a region's foreground mask
  * Replace the background with the mean input so that background regions are zero after mean subs
* **full+fg**: concatenates the full and fg features

![image](https://user-images.githubusercontent.com/46519736/71569862-47f88b80-2b15-11ea-830a-e973a6e4b94f.png)

* Layer fc_6 always outperforms fc_7
* fg outpuerforms full: masked region shape provides a stronger signal
* full + fg: Best result by a margin of 4.2%
  * The context provided by the full features is highly informative even given the fg features
